cache plan if no plan specified

jatin837 opened the pull request:
update plan if passed as argument

select plan if no cache present and no args given and update cache accordingly

viferga commented:
The implementation looks good but as @Creatoon suggested in the Discord group, we can do this more scalable. It should be able to index plans per project.

Creatoon commented:
### Considerations
1) Looking at this pr, ig we are caching the plans in the ram and once the process stops, the whole cache will be flushed and that will not be good.
2) For caching other stuff rather than plans, we have to change the structure of the cache object in the code.

### Proposed Soln
1) We should cache the preferences of the project in ```.metacall/deploy/deployments.ini``` file and the structure looks like this:- 

```ini
[home/raj/desktop/authMiddleware]
PLAN=essential
KEY1=value

[home/raj/desktop/projects/time-web-app]
PLAN=premium
KEY1=value

[https://github.com/user/repo]
PLAN=standard
KEY1=value
```

viferga commented:
Would you use a different file then? I think the structure is good.

jatin837 commented:
in my implementation, I stored cache as key-value pair
where key => project directory
and value => previous plan

here, i'm storing cache as record<string, string>

jatin837 commented:
Other than plan, what else can be cached?

jatin837 commented:
This time, I changed the file name and the format of cache file.
all cache is stored in 
> metacall/deploy/cache.ini

also, this time I've added the an example cache file.
I'll write a test case for cache using that cache.ini file

Creatoon commented on src/config.ts:
```diff
@@ -21,18 +21,49 @@ export const Config = z.object({
+	path = cacheFilePath
+) => {
+	const cache = await loadCache(path);
+	cache[key] = { plan: val };
```
@viferga wdyt?
can you change the caching preferences to general key-value pairs, this is again the same stuff, CLI should not be limited to caching plans, yes, I know there are not many things right now to cache but in the future, we may, so changes I could suggest is.

Pass key-value pair for caching.
for the above example ```key=plan``` and ```val=planValue```

jatin837 replied:
I'm doing what you're describing. Here cache is a record of a project directory as key and an object as value.

That object (value) can contain many things.

But so far, it only contains plan as key.

Creatoon replied:
aren't you hard coding only plan?

jatin837 replied:
> @viferga wdyt? Great, Can you just change the caching preferences to general key-value pairs, this is again the same stuff, CLI should not be limited to caching plans, yes, I know there are not many things right now to cache but in the future, we may, so changes I could suggest is.
> 
> Pass key-value pair for caching. for the above example `key=plan` and `val=planValue`

I guess we can do that, but then we might loose type safety.

We can make keys optional but still, we need to know what keys are we taking about

Auth token, name, password, e-mail address are stored in base config, so no point of including them in cache.

Creatoon replied:
Yeah exactly, I guess creation an interface might solve the problem, like it will work as schema for cache.ini, wdyt?

jatin837 replied:
Well, it doesn't look like schema is the problem here.

Zod is fine here as it provides optional keys. But what keys? We still need them in interfaces as well.

The problem is what keys do we need other than plan.

Can we enumerate what else to cache?🤔

Creatoon replied:
That's I am saying from starting that for now plan is the only thing to be cached, but previously we were also caching env vara, but as @viferga said we will cache it on db then that's also not an option here but we might encounter some extra preferences to be cached in the future, so we don't have to tweak the code much that's what I meant by scalable, but anyways let's wait what @viferga says.
